analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -35,8 +35,22 @@ public sealed record SymbolicValue
         public static readonly SymbolicValue True = NotNull.WithConstraint(BoolConstraint.True);
         public static readonly SymbolicValue False = NotNull.WithConstraint(BoolConstraint.False);
 
+        private readonly ImmutableDictionary<Type, SymbolicConstraint> constraints = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private int? constraintsHashCode;
+
         // SymbolicValue can have only one constraint instance of specific type at a time
-        private ImmutableDictionary<Type, SymbolicConstraint> Constraints { get; init; } = ImmutableDictionary<Type, SymbolicConstraint>.Empty;
+        private ImmutableDictionary<Type, SymbolicConstraint> Constraints
+        {
+            get => constraints;
+            init
+            {
+                if (constraints != value)
+                {
+                    constraints = value;
+                    constraintsHashCode = null;
+                }
+            }
+        }
```

Comment:
There is also 
https://github.com/SonarSource/sonar-dotnet/blob/f3c2ea1f0dcbdf8731ecd081325a558932c91561/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs#L138-L143

and I want to use the same pattern there as well. There it is more important to reset the cached hashcode in the init setter. There we can do this, to make the declaration more compact:

```cs
private readonly ImmutableDictionary<IOperation, SymbolicValue> operationValue;
private int? operationValueHashCode;

private ImmutableDictionary<IOperation, SymbolicValue> OperationValue
{
    get => operationValue;
    init => SetHashCodeCachedField(value, ref operationValue, ref operationValueHashCode);
}

private static void SetHashCodeCachedField<T>(T value, ref T backingField, ref int? backingFieldHashCode) where T: class
{
    if (value == backingField)
    {
        backingField = value;
        backingFieldHashCode = null;
    }
}
```

Reply:
The `init` and `ref` trick sounds good in principle. 

`backingFieldHashCode` should not be a parameter, the method can just reset it directly.